torch/distributed/_shard/checkpoint/metadata.py

```diff
@@ -0,0 +1,79 @@
+import io
+import pickle
+from dataclasses import dataclass
+from typing import Dict, List, Tuple, Union
+
+import torch
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+    ShardedTensorMetadata,
+    ShardMetadata,
+)
+
+TENSOR_TYPE = Union[torch.Tensor, ShardedTensor]
+
+@dataclass
+class ShardStorageMetadata:
+    shard_metadata: ShardMetadata
+    # Unique identifier for this particular Shard
+    storage_key: str
+    # Length in bytes for this shard
+    length: int
+
+
+# Metadata for each param.
+@dataclass
+class ShardedTensorStorageMetadata:
+    # Metadata for the sharded tensor itself
+    tensor_metadata: ShardedTensorMetadata
+    # Storage info for each Shard. There's no ordering requirement for this list.
+    storage_metadata: List[ShardStorageMetadata]
+
+
+@dataclass
+class TensorStorageMetadata:
+    # Unique indentifier for this tensor
+    storage_key: str
+
+    # Tensor sizes
+    size: Tuple[int, ...]
+
+@dataclass
+class Metadata:
+    # Metadata for the state dict.
+    # This includes the MD for Tensors and ShardedTensors. ByteIO objects are skipped
+    state_dict_metadata: Dict[str, Union[ShardedTensorStorageMetadata, TensorStorageMetadata]]
+
+    def __getstate__(self) -> bytes:
+        serialized = pickle.dumps(self.state_dict_metadata)
+        return serialized
+
+    def __setstate__(self, state: bytes) -> None:
+        self.state_dict_metadata = pickle.loads(state)
+
+
+@dataclass
+class BytesWriteRequest:
+    bytes: io.BytesIO
+    storage_key: str
+
+
+@dataclass
+class BytesReadRequest:
+    bytes: io.BytesIO
+    storage_key: str
```

Comment:
Good point. IIRC for now there is no usage for non-tensor resharding. But might be helpful for feature like "checkpoint per-rank data loaders buffer". In practical, dataloader has a buffer to store prefetched batches but not trained yet. When saving data loader checkpoint, we need to drain their buffers, otherwise we will loss batches in buffer when resume it (data loss).

If we also want to save these "per rank dataloader buffer" in checkpoint as bytes, a reshardable BytesReadRequest is helpful when number of dataloaders (== number of trainers) change after resume

Reply:
That's sounds like an important scenario to cover, but quite far from the current design we currently have.
I'm taking notes of this so we can tinker with it.